File: packages/tracelet_platform_interface/test/algorithms/trip_manager_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:tracelet_platform_interface/tracelet_platform_interface.dart';

void main() async {
  await RustLib.init();
  group('TripManager', () {
    late TripManager tripManager;
    late List<Map<String, Object?>> tripEvents;

    setUp(() {
      tripManager = TripManager();
      tripEvents = <Map<String, Object?>>[];
      tripManager.onTripEnd = (data) => tripEvents.add(data);
    });

    test('starts inactive', () {
      expect(tripManager.isTripActive, isFalse);
    });

    test('starts trip on isMoving=true', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 37.4220,
        longitude: -122.0840,
        timestamp: '2024-01-01T00:00:00Z',
      );

      expect(tripManager.isTripActive, isTrue);
      expect(tripEvents, isEmpty); // No trip end yet
    });

    test('ends trip on isMoving=false and emits event', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 37.4220,
        longitude: -122.0840,
        timestamp: '2024-01-01T00:00:00Z',
      );

      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 37.4230,
        longitude: -122.0830,
        timestamp: '2024-01-01T00:05:00Z',
      );

      expect(tripManager.isTripActive, isFalse);
      expect(tripEvents.length, 1);

      final trip = tripEvents.first;
      expect(trip['isMoving'], false);
      expect(trip['distance'], isA<double>());
      expect(trip['duration'], isA<double>());
      expect(trip['startLocation'], isA<Map>());
      expect(trip['stopLocation'], isA<Map>());
      expect(trip['waypoints'], isA<List>());
    });

    test('records start location in trip event', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 37.4220,
        longitude: -122.0840,
      );
      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 37.4230,
        longitude: -122.0830,
      );

      final startLoc =
          tripEvents.first['startLocation']! as Map<String, Object?>;
      expect(startLoc['latitude'], 37.4220);
      expect(startLoc['longitude'], -122.0840);

      final stopLoc = tripEvents.first['stopLocation']! as Map<String, Object?>;
      expect(stopLoc['latitude'], 37.4230);
      expect(stopLoc['longitude'], -122.0830);
    });

    test('accumulates waypoints during trip', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 37.4220,
        longitude: -122.0840,
      );

      // Feed locations during trip
      tripManager.onLocationReceived(
        latitude: 37.4221,
        longitude: -122.0841,
        timestamp: 'ts1',
      );
      tripManager.onLocationReceived(
        latitude: 37.4222,
        longitude: -122.0842,
        timestamp: 'ts2',
      );
      tripManager.onLocationReceived(
        latitude: 37.4223,
        longitude: -122.0843,
        timestamp: 'ts3',
      );

      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 37.4224,
        longitude: -122.0844,
      );

      final waypoints =
          tripEvents.first['waypoints']! as List<Map<String, Object?>>;
      // start waypoint + 3 locations + stop waypoint = 5
      expect(waypoints.length, 5);
    });

    test('ignores locations when no trip active', () {
      tripManager.onLocationReceived(latitude: 37.4220, longitude: -122.0840);

      // Start and immediately stop
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 11,
        longitude: 21,
      );

      final waypoints =
          tripEvents.first['waypoints']! as List<Map<String, Object?>>;
      // Only start + stop = 2
      expect(waypoints.length, 2);
    });

    test('calculates distance using Haversine', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 37,
        longitude: -122,
      );

      // Move ~111 km north (1 degree latitude)
      tripManager.onLocationReceived(latitude: 38, longitude: -122);

      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 38,
        longitude: -122,
      );

      final distance = tripEvents.first['distance']! as double;
      // 1 degree latitude ≈ 111,195 meters (±500m tolerance)
      expect(distance, closeTo(111195, 500));
    });

    test('reset clears state', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      expect(tripManager.isTripActive, isTrue);

      tripManager.reset();

      expect(tripManager.isTripActive, isFalse);
      // No trip end event should be fired on reset
      expect(tripEvents, isEmpty);
    });

    test('duplicate isMoving=true does not restart trip', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      tripManager.onLocationReceived(latitude: 11, longitude: 21);
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 12,
        longitude: 22,
      );

      // Should still be the same trip
      expect(tripManager.isTripActive, isTrue);

      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 13,
        longitude: 23,
      );

      expect(tripEvents.length, 1);
      // Waypoints: start(10,20) + location(11,21) + stop(13,23) = 3
      final waypoints =
          tripEvents.first['waypoints']! as List<Map<String, Object?>>;
      expect(waypoints.length, 3);
    });

    test('duplicate isMoving=false does not emit multiple trips', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 11,
        longitude: 21,
      );
      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 12,
        longitude: 22,
      );

      expect(tripEvents.length, 1);
    });

    test('onTripEnd not called if callback is null', () {
      tripManager.onTripEnd = null;

      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 11,
        longitude: 21,
      );

      // Should not throw
      expect(tripEvents, isEmpty);
    });

    test('trip with null start/stop coordinates', () {
      tripManager.onMotionStateChanged(isMoving: true);
      tripManager.onMotionStateChanged(isMoving: false);

      expect(tripEvents.length, 1);
      final startLoc =
          tripEvents.first['startLocation']! as Map<String, Object?>;
      expect(startLoc, isEmpty);
    });

    test('waypoint cap evicts oldest entries', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 0,
        longitude: 0,
      );

      // Feed 5001 locations (exceeds _maxWaypoints = 5000).
      // Start waypoint counts as 1, so we need 5000 more to trigger eviction.
      for (var i = 1; i <= 5001; i++) {
        tripManager.onLocationReceived(
          latitude: i.toDouble(),
          longitude: i.toDouble(),
          timestamp: 'ts$i',
        );
      }

      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 9999,
        longitude: 9999,
      );

      final waypoints =
          tripEvents.first['waypoints']! as List<Map<String, Object?>>;
      // Should be capped at 5000 (oldest evicted) + 1 stop waypoint
      // = 5000 total (cap enforced on add, stop adds one more but
      // onLocationReceived already added 5001 which was capped to 5000,
      // then endTrip adds 1 more which triggers another eviction = 5000).
      expect(waypoints.length, lessThanOrEqualTo(5001));
      expect(waypoints.length, greaterThan(4999));
    });

    test('accumulates distance across multiple locations', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 0,
        longitude: 0,
      );

      // Move along a straight line: 0,0 → 0,1 → 0,2
      tripManager.onLocationReceived(latitude: 0, longitude: 1);
      tripManager.onLocationReceived(latitude: 0, longitude: 2);

      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 0,
        longitude: 2,
      );

      final distance = tripEvents.first['distance']! as double;
      // 2 degrees of longitude at equator ≈ 222,390m
      expect(distance, closeTo(222390, 1000));
    });

    test('duration is positive for real trip', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      // Duration depends on wall clock so just ensure it's >= 0
      tripManager.onMotionStateChanged(
        isMoving: false,
        latitude: 11,
        longitude: 21,
      );

      final duration = tripEvents.first['duration']! as double;
      expect(duration, greaterThanOrEqualTo(0));
    });

    test('consecutive start-stop cycles produce separate trips', () {
      for (var i = 0; i < 3; i++) {
        tripManager.onMotionStateChanged(
          isMoving: true,
          latitude: i.toDouble(),
          longitude: i.toDouble(),
        );
        tripManager.onMotionStateChanged(
          isMoving: false,
          latitude: i + 1.0,
          longitude: i + 1.0,
        );
      }

      expect(tripEvents.length, 3);
    });

    test('reset during active trip does not emit event', () {
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 10,
        longitude: 20,
      );
      tripManager.onLocationReceived(latitude: 11, longitude: 21);
      tripManager.reset();

      expect(tripManager.isTripActive, isFalse);
      expect(tripEvents, isEmpty);

      // Can start a new trip after reset
      tripManager.onMotionStateChanged(
        isMoving: true,
        latitude: 30,
        longitude: 40,
      );
      expect(tripManager.isTripActive, isTrue);
    });
  });
}

File: sdk/android/tracelet-sdk/src/test/kotlin/com/ikolvi/tracelet/sdk/algorithm/TripManagerTest.kt
package com.ikolvi.tracelet.sdk.algorithm

import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Unit tests for [TripManager].
 */
class TripManagerTest {

    private lateinit var tripManager: TripManager
    private var lastTripData: Map<String, Any?>? = null

    @Before
    fun setUp() {
        tripManager = TripManager()
        lastTripData = null
        tripManager.onTripEnd = { data -> lastTripData = data }
    }

    @Test
    fun `trip starts on moving transition`() {
        assertFalse(tripManager.isTripActive)
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        assertTrue(tripManager.isTripActive)
        assertNull(lastTripData) // Trip hasn't ended yet
    }

    @Test
    fun `trip ends on stationary transition`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.43, longitude = -122.07)

        assertFalse(tripManager.isTripActive)
        assertNotNull(lastTripData)
        assertEquals(false, lastTripData!!["isMoving"])
    }

    @Test
    fun `trip collects start and stop locations`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.43, longitude = -122.07)

        val start = lastTripData!!["startLocation"] as Map<*, *>
        val stop = lastTripData!!["stopLocation"] as Map<*, *>
        assertEquals(37.42, start["latitude"])
        assertEquals(-122.08, start["longitude"])
        assertEquals(37.43, stop["latitude"])
        assertEquals(-122.07, stop["longitude"])
    }

    @Test
    fun `trip accumulates distance from waypoints`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onLocationReceived(37.421, -122.079)
        tripManager.onLocationReceived(37.422, -122.078)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.423, longitude = -122.077)

        val distance = lastTripData!!["distance"] as Double
        assertTrue(distance > 0, "Distance should be positive: $distance")
    }

    @Test
    fun `trip records duration`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        // Small delay is inherent (system time), but duration should be >= 0
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.43, longitude = -122.07)

        val duration = lastTripData!!["duration"] as Double
        assertTrue(duration >= 0)
    }

    @Test
    fun `waypoints recorded during trip`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onLocationReceived(37.421, -122.079, "ts1")
        tripManager.onLocationReceived(37.422, -122.078, "ts2")
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.423, longitude = -122.077)

        @Suppress("UNCHECKED_CAST")
        val waypoints = lastTripData!!["waypoints"] as List<Map<String, Any?>>
        // Start waypoint + 2 location + end stop waypoint
        assertTrue(waypoints.size >= 3, "Expected at least 3 waypoints, got ${waypoints.size}")
    }

    @Test
    fun `locations ignored when trip not active`() {
        tripManager.onLocationReceived(37.42, -122.08) // No trip started
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.42, longitude = -122.08)

        @Suppress("UNCHECKED_CAST")
        val waypoints = lastTripData!!["waypoints"] as List<Map<String, Any?>>
        // Only start and end waypoints, no pre-trip locations
        assertTrue(waypoints.size <= 2)
    }

    @Test
    fun `duplicate moving events do not restart trip`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onLocationReceived(37.421, -122.079)
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.422, longitude = -122.078)
        // Should still be same trip
        assertTrue(tripManager.isTripActive)
        assertNull(lastTripData) // No trip ended
    }

    @Test
    fun `reset clears trip state`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        assertTrue(tripManager.isTripActive)

        tripManager.reset()
        assertFalse(tripManager.isTripActive)
    }

    @Test
    fun `reset during trip does not fire onTripEnd`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.reset()
        assertNull(lastTripData)
    }

    @Test
    fun `multiple consecutive trips`() {
        // First trip
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.43, longitude = -122.07)
        val firstTrip = lastTripData
        assertNotNull(firstTrip)

        // Second trip
        lastTripData = null
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.44, longitude = -122.06)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.45, longitude = -122.05)
        val secondTrip = lastTripData
        assertNotNull(secondTrip)

        // Verify they have different start locations
        val start1 = firstTrip["startLocation"] as Map<*, *>
        val start2 = secondTrip["startLocation"] as Map<*, *>
        assertEquals(37.42, start1["latitude"])
        assertEquals(37.44, start2["latitude"])
    }

    @Test
    fun `trip with null coordinates`() {
        tripManager.onMotionStateChanged(isMoving = true)
        assertTrue(tripManager.isTripActive)
        tripManager.onMotionStateChanged(isMoving = false)
        assertFalse(tripManager.isTripActive)
        assertNotNull(lastTripData)
        assertEquals(0.0, lastTripData!!["distance"])
    }

    @Test
    fun `average speed calculated correctly`() {
        tripManager.onMotionStateChanged(isMoving = true, latitude = 37.42, longitude = -122.08)
        tripManager.onLocationReceived(37.421, -122.079)
        tripManager.onMotionStateChanged(isMoving = false, latitude = 37.422, longitude = -122.078)

        val distance = lastTripData!!["distance"] as Double
        val duration = lastTripData!!["duration"] as Double
        if (duration > 0) {
            val avgSpeed = distance / duration
            assertTrue(avgSpeed >= 0)
        }
    }
}
